[opencv2] build failure

**Host Environment**
- OS: Windows 10 21H2
- Compiler: MSVC 19.29.30038.1

**To Reproduce**
Steps to reproduce the behavior:
```cmd
cd C:\C++\vcpkg
.\vcpkg install opencv2
```

**Failure logs** 
```
-- Found external ninja('1.10.2').
-- Configuring x64-windows
CMake Error at scripts/cmake/vcpkg_execute_required_process.cmake:127 (message):
    Command failed: ninja -v
    Working Directory: C:/C++/vcpkg/buildtrees/opencv2/x64-windows-rel/vcpkg-parallel-configure
    Error code: 1
    See logs for more information:
      C:\C++\vcpkg\buildtrees\opencv2\config-x64-windows-out.log

Call Stack (most recent call first):
  installed/x64-windows/share/vcpkg-cmake/vcpkg_cmake_configure.cmake:364 (vcpkg_execute_required_process)
  ports/opencv2/portfile.cmake:61 (vcpkg_cmake_configure)
  scripts/ports.cmake:142 (include)


Error: Building package opencv2:x64-windows failed with: BUILD_FAILED
Please ensure you're using the latest portfiles with `git pull` and `.\vcpkg update`, then
submit an issue at https://github.com/Microsoft/vcpkg/issues including:
  package: opencv2[core,eigen,jpeg,png,tiff]:x64-windows -> 2.4.13.7#6
  vcpkg version: 2021-11-02-af04ebf6274fd6f7a941bff4662b3955c64f6f42
  vcpkg-tool version: 7375bb408 2021-11-03 (71 minutes ago)
```
config-x64-windows-out.log:
```
-- Detecting C compile features - done
CMake Error at cmake/OpenCVUtils.cmake:90 (if):
  if given arguments:

    "C:/C++/vcpkg/buildtrees/opencv2/x64-windows-dbg" "MATCHES" "^C:/C++/vcpkg/buildtrees/opencv2/src/2.4.13.7-64eb3208d2.clean" "OR" "C:/C++/vcpkg/buildtrees/opencv2/x64-windows-dbg" "MATCHES" "^C:/C++/vcpkg/buildtrees/opencv2/x64-windows-dbg"

  Regular expression
  "^C:/C++/vcpkg/buildtrees/opencv2/src/2.4.13.7-64eb3208d2.clean" cannot
  compile
Call Stack (most recent call first):
  CMakeLists.txt:351 (ocv_include_directories)


-- Configuring incomplete, errors occurred!
See also "C:/C++/vcpkg/buildtrees/opencv2/x64-windows-dbg/CMakeFiles/CMakeOutput.log".
RegularExpression::compile(): Nested *?+.
RegularExpression::compile(): Error in compile.
```
OpenCVUtils.cmake:
```cmake
# adds include directories in such a way that directories from the OpenCV source tree go first
function(ocv_include_directories)
  set(__add_before "")
  foreach(dir ${ARGN})
    get_filename_component(__abs_dir "${dir}" ABSOLUTE)
    if("${__abs_dir}" MATCHES "^${OpenCV_SOURCE_DIR}" OR "${__abs_dir}" MATCHES "^${OpenCV_BINARY_DIR}")
      list(APPEND __add_before "${dir}")
    elseif(CMAKE_COMPILER_IS_GNUCXX AND NOT CMAKE_CXX_COMPILER_VERSION VERSION_LESS "6.0" AND
           dir MATCHES "/usr/include$")
      # workaround for GCC 6.x bug
    else()
      include_directories(AFTER SYSTEM "${dir}")
    endif()
  endforeach()
  include_directories(BEFORE ${__add_before})
endfunction()
```

**Additional context**
One solution (from opencv4/[OpenCVUtils.cmake](https://github.com/opencv/opencv/blob/c703f1eed6c3da4f14b37cc151e03463bb4ff54f/cmake/OpenCVUtils.cmake)):
```cmake
# adds include directories in such a way that directories from the OpenCV source tree go first
function(ocv_include_directories)
  ocv_debug_message("ocv_include_directories( ${ARGN} )")
  set(__add_before "")
  foreach(dir ${ARGN})
    ocv_is_opencv_directory(__is_opencv_dir "${dir}")
    if(__is_opencv_dir)
      list(APPEND __add_before "${dir}")
    elseif(((CV_GCC AND NOT CMAKE_CXX_COMPILER_VERSION VERSION_LESS "6.0") OR CV_CLANG) AND
           dir MATCHES "/usr/include$")
      # workaround for GCC 6.x bug
    else()
      include_directories(AFTER SYSTEM "${dir}")
    endif()
  endforeach()
  include_directories(BEFORE ${__add_before})
endfunction()

function(ocv_is_opencv_directory result_var dir)
  set(result FALSE)
  foreach(parent ${OpenCV_SOURCE_DIR} ${OpenCV_BINARY_DIR} ${OPENCV_EXTRA_MODULES_PATH})
    ocv_is_subdir(result "${parent}" "${dir}")
    if(result)
      break()
    endif()
  endforeach()
  set(${result_var} ${result} PARENT_SCOPE)
endfunction()

# check if "sub" (file or dir) is below "dir"
function(ocv_is_subdir res dir sub )
  get_filename_component(dir "${dir}" ABSOLUTE)
  get_filename_component(sub "${sub}" ABSOLUTE)
  file(TO_CMAKE_PATH "${dir}" dir)
  file(TO_CMAKE_PATH "${sub}" sub)
  set(dir "${dir}/")
  string(LENGTH "${dir}" len)
  string(LENGTH "${sub}" len_sub)
  if(NOT len GREATER len_sub)
    string(SUBSTRING "${sub}" 0 ${len} prefix)
  endif()
  if(prefix AND prefix STREQUAL dir)
    set(${res} TRUE PARENT_SCOPE)
  else()
    set(${res} FALSE PARENT_SCOPE)
  endif()
endfunction()
```

--- a/ports/opencv2/fix-path-contains-++-error.patch
+++ b/ports/opencv2/fix-path-contains-++-error.patch
@@ -1,0 +1,49 @@
+diff --git a/cmake/OpenCVUtils.cmake b/cmake/OpenCVUtils.cmake
+index 72df4d4..7d45974 100644
+--- a/cmake/OpenCVUtils.cmake
++++ b/cmake/OpenCVUtils.cmake
+@@ -82,12 +82,42 @@ macro(ocv_check_environment_variables)
+   endforeach()
+ endmacro()
+ 
++# check if "sub" (file or dir) is below "dir"
++function(is_subdir res dir sub )
++  get_filename_component(dir "${dir}" ABSOLUTE)
++  get_filename_component(sub "${sub}" ABSOLUTE)
++  file(TO_CMAKE_PATH "${dir}" dir)
++  file(TO_CMAKE_PATH "${sub}" sub)
++  set(dir "${dir}/")
++  string(LENGTH "${dir}" len)
++  string(LENGTH "${sub}" len_sub)
++  if(NOT len GREATER len_sub)
++    string(SUBSTRING "${sub}" 0 ${len} prefix)
++  endif()
++  if(prefix AND prefix STREQUAL dir)
++    set(${res} TRUE PARENT_SCOPE)
++  else()
++    set(${res} FALSE PARENT_SCOPE)
++  endif()
++endfunction()
++
++function(ocv_is_opencv_directory result_var dir)
++  set(result FALSE)
++  foreach(parent ${OpenCV_SOURCE_DIR} ${OpenCV_BINARY_DIR} ${OPENCV_EXTRA_MODULES_PATH})
++    is_subdir(result "${parent}" "${dir}")
++    if(result)
++      break()
++    endif()
++  endforeach()
++  set(${result_var} ${result} PARENT_SCOPE)
++endfunction()
++
+ # adds include directories in such a way that directories from the OpenCV source tree go first
+ function(ocv_include_directories)
+   set(__add_before "")
+   foreach(dir ${ARGN})
+-    get_filename_component(__abs_dir "${dir}" ABSOLUTE)
+-    if("${__abs_dir}" MATCHES "^${OpenCV_SOURCE_DIR}" OR "${__abs_dir}" MATCHES "^${OpenCV_BINARY_DIR}")
++    ocv_is_opencv_directory(__is_opencv_dir "${dir}")
++    if(__is_opencv_dir)
+       list(APPEND __add_before "${dir}")
+     elseif(CMAKE_COMPILER_IS_GNUCXX AND NOT CMAKE_CXX_COMPILER_VERSION VERSION_LESS "6.0" AND
+            dir MATCHES "/usr/include$")

--- a/ports/opencv2/portfile.cmake
+++ b/ports/opencv2/portfile.cmake
@@ -26,6 +26,7 @@ vcpkg_from_github(
       0003-force-package-requirements.patch
       0004-add-ffmpeg-missing-defines.patch
       0005-fix-cuda.patch
+      fix-path-contains-++-error.patch
 )
 
 file(REMOVE "${SOURCE_PATH}/cmake/FindCUDA.cmake")

--- a/ports/opencv2/vcpkg.json
+++ b/ports/opencv2/vcpkg.json
@@ -1,7 +1,7 @@
 {
   "name": "opencv2",
   "version": "2.4.13.7",
-  "port-version": 6,
+  "port-version": 7,
   "description": "computer vision library",
   "homepage": "https://github.com/opencv/opencv",
   "dependencies": [

--- a/versions/baseline.json
+++ b/versions/baseline.json
@@ -4882,7 +4882,7 @@
     },
     "opencv2": {
       "baseline": "2.4.13.7",
-      "port-version": 6
+      "port-version": 7
     },
     "opencv3": {
       "baseline": "3.4.15",

--- a/versions/o-/opencv2.json
+++ b/versions/o-/opencv2.json
@@ -1,6 +1,11 @@
 {
   "versions": [
     {
+      "git-tree": "d6ccb4fb79829bcf58dd79950529d7086d6b277c",
+      "version": "2.4.13.7",
+      "port-version": 7
+    },
+    {
       "git-tree": "e85a45aa17a47da9f965b93cb0c6fc8273a7ca04",
       "version": "2.4.13.7",
       "port-version": 6

